File: meetings/5/JRubyResolvingTests/test_runner.rb
require 'java'
# [TODO] make path
require 'C:\Users\Alexey\Desktop\Kotlin\1_prepare\JRubyResolvingTests\out\artifacts\javaTestCore_jar\javaTestCore.jar'

module JavaTestCoreUtils
  include_package "resolving.tests"
end

def assertEquals(value, expected)
  was = false
  if expected.kind_of?(Array);
    for current in expected;
      if current == value;
        was = true
        break
      end
    end
  else
    if expected == value;
      was = true
    end
  end
  unless was;
    raise "This is wrong, actual = #{value}" unless expected == value
  end
end

class TestRunner
end


br = JavaTestCoreUtils::BasicRouter.new

x = 1
puts "Обычное разрешение"
assertEquals(br.method_(x), [2, 3])
# 36: warning: ambiguous Java methods found, using method_(int/short)
assertEquals(br.method_("one"), 0)
assertEquals(br.method_(29), [2, 3])
assertEquals(br.method_(Integer 29), [2, 3])
assertEquals(br.method_(2.9), [2, 3, 4]) #41: warning: ambiguous Java methods found, using method_(short)
#; WAT????? double!!!!


puts "Integer"
assertEquals(br.method_IntDyn(x), 0)

puts ("Supplier")

usualSupplier = JavaTestCoreUtils::Supplier.new
assertEquals(br.method_(usualSupplier.getInt), [2, 3])
assertEquals(br.method_((Integer usualSupplier.getInt)), [2, 3])
assertEquals(br.method_(usualSupplier.getString), 0)
assertEquals(br.method_(usualSupplier.getInteger), [2, 3])
assertEquals(br.method_(usualSupplier.getDouble), [2, 3, 4])

puts ("Supplier with trait")
traitSupplier = JavaTestCoreUtils::DefaultInterface_Trait_Supplier.new
assertEquals(br.method_(traitSupplier.getInt), [2, 3]) #
assertEquals(br.method_((Integer traitSupplier.getInt)), [2, 3]) #
assertEquals(br.method_(traitSupplier.getString), 0)
assertEquals(br.method_(traitSupplier.getInteger), [2, 3]) #
assertEquals(br.method_(traitSupplier.getDouble), [2, 3, 4]) #

puts ("Inheritance without integer")
assertEquals(br.methodNoInteger__(traitSupplier.getBase, traitSupplier.getInt), 0)
assertEquals(br.methodNoInteger__(traitSupplier.getBase, traitSupplier.getInteger), 0)
assertEquals(br.methodNoInteger__(traitSupplier.getDerived, traitSupplier.getInteger), 3) #
assertEquals(br.methodNoInteger__(traitSupplier.getDerived2lvl, traitSupplier.getInteger), 5)

puts ("Inheritance with integer")
assertEquals(br.method__(traitSupplier.getBase, traitSupplier.getInt), 0)
assertEquals(br.method__(traitSupplier.getBase, traitSupplier.getInteger), 0)
assertEquals(br.method__(traitSupplier.getDerived, traitSupplier.getInteger), 3) #

puts ("Supplier with Null")
assertEquals(br.method_(traitSupplier.getNull), [0, 4, 6]) ##77: warning: ambiguous Java methods found, using method_(java.lang.String)
assertEquals(br.method_(traitSupplier.getDerivedNull), [0, 4, 6]) #
assertEquals(br.method_(traitSupplier.getDerivedTypedNull), [0, 4, 6]) #

puts ("Ambiguous Interface Router")
ar = JavaTestCoreUtils::AmbiguousRouter.new
is = JavaTestCoreUtils::InterfaceSupplier.new
assertEquals(ar.method_IP12(is.getInterfaceProvider), 1)
assertEquals(ar.method_I12(is.getInterfaceProvider), 3)##85: warning: ambiguous Java methods found, using method_I12(resolving.tests.I2)

puts ("No Default Router ON JVM")

File: meetings/5/Clojure/src/TestRunner.clj
(ns TestRunner
    (:import (resolving.tests BasicRouter Supplier DefaultInterface_Trait_Supplier Derived AmbiguousRouter InterfaceSupplier DefaultRouter)))

(print (.toUpperCase "fred"))

(def br (new BasicRouter))

(println "Обычное разрешение")

(def xdyn 1)
(def x (int 1))

(assert (= 4 (. br method_ x)))
(assert (= 0 (. br method_ (new String "one"))))
(assert (= 1 (. br method_ 29)))
(assert (= 4 (. br method_ (int 29))))
(assert (= 4 (. br method_ (new Integer 29))))
(assert (= 1 (. br method_ 2.9)))
(assert (= 1 (. br method_ xdyn)))

(println "Integer")

(assert (= 1 (. br method_IntDyn xdyn)))
(assert (= 1 (. br method_IntDyn 1)))
(assert (= 1 (. br method_IntDyn (int 1))))



(println "Supplier")

(def usualSupplier (new Supplier))

(assert (= 4 (. br method_ (. usualSupplier getInt))))
(assert (= 4 (. br method_ (int (. usualSupplier getInt)))))
(assert (= 0 (. br method_ (. usualSupplier getString))))
(assert (= 4 (. br method_ (. usualSupplier getInteger))))
(assert (= 1 (. br method_ (. usualSupplier getDouble))))

(println "Default Interface Suppplier")
(def traitSupplier (new DefaultInterface_Trait_Supplier))

(assert (= 4 (. br method_ (. traitSupplier getInt))))
(assert (= 4 (. br method_ (int (. traitSupplier getInt)))))
(assert (= 0 (. br method_ (. traitSupplier getString))))
(assert (= 4 (. br method_ (. traitSupplier getInteger))))
(assert (= 1 (. br method_ (. traitSupplier getDouble))))


(println "Inheritance without integer")
(assert (= 0 (. br methodNoInteger__ (. traitSupplier getBase) (. traitSupplier getInt))))
(assert (= 0 (. br methodNoInteger__ (. traitSupplier getBase) (. traitSupplier getInteger))))
(assert (= 1 (. br methodNoInteger__ (. traitSupplier getDerived) (. traitSupplier getInteger))))
(assert (= 5 (. br methodNoInteger__ (. traitSupplier getDerived2lvl) (. traitSupplier getInteger))))


(println "Inheritance with integer")
(assert (= 0 (. br method__ (. traitSupplier getBase) (. traitSupplier getInt))))
(assert (= 0 (. br method__ (. traitSupplier getBase) (. traitSupplier getInteger))))
(assert (= 4 (. br method__ (. traitSupplier getDerived) (. traitSupplier getInteger))))


(println "Supplier with Null")

(println (. br method_ (. traitSupplier getNull)))
(println (. br method_ (. traitSupplier getDerivedNull)))
(println (. br method_ (cast Derived (. traitSupplier getDerivedNull))))
(println (. br method_ (. traitSupplier getDerivedTypedNull)))

(println "6 or 0 or 4")
; (assert (= 4 (. br method_ (. traitSupplier getNull))))
; (assert (= 4 (. br method_ (. traitSupplier getDerivedNull))))
; (assert (= 4 (. br method_ (cast Derived (. traitSupplier getDerivedNull)))))
; (assert (= 4 (. br method_ (. traitSupplier getDerivedTypedNull))))


(println "Ambiguous Interface Router")
(def ar (new AmbiguousRouter))
(def is (new InterfaceSupplier))


(assert (= 1 (. ar method_IP12 (. is getInterfaceProvider))))
(assert (= 2 (. ar method_I12 (. is getInterfaceProvider))))
; !!!

(println "Default Router")
(println "dont work with java")
(def dr (new DefaultRouter))
(assert (= 1 (. dr method__ (. traitSupplier getDerived))))
; (assert (= 2 (. dr method__ (. traitSupplier getBase))))
